pkg/network/ovn_kubernetes.go

```diff
@@ -1840,3 +1731,164 @@ func isIPSecEnabledInPod(pod v1.PodTemplateSpec, containerName string) bool {
+			}
+			if err := validateOVNKubernetesSubnet("v6InternalTransitSwitchSubnet", oc.IPv6.InternalTransitSwitchSubnet, &pool, conf.ClusterNetwork); err != nil {
+				out = append(out, err)
+			}
+		}
+	}
+
+	// oc.V4InternalSubnet and oc.IPv4.InternalJoinSubnet must be same if both are present
+	if oc.V4InternalSubnet != "" && oc.IPv4 != nil {
+		if oc.V4InternalSubnet != oc.IPv4.InternalJoinSubnet {
+			out = append(out, errors.Errorf("v4InternalSubnet will be deprecated soon, until then it must be same as v4InternalJoinSubnet %s ", oc.IPv4.InternalJoinSubnet))
+		}
+	}
+
+	// oc.V6InternalSubnet and oc.IPv6.InternalJoinSubnet must be same if both are present
+	if oc.V6InternalSubnet != "" && oc.IPv6 != nil {
+		if oc.V6InternalSubnet != oc.IPv6.InternalJoinSubnet {
+			out = append(out, errors.Errorf("v6InternalSubnet will be deprecated soon, until then it must be same as v6InternalJoinSubnet %s ", oc.IPv6.InternalJoinSubnet))
+		}
+	}
+
+	// Gateway Configurable Subnet Checks
+	//Validate whether masquerade CIDR is from same IP family as clusterNetwork.
+	if oc.GatewayConfig != nil {
+		if oc.GatewayConfig.IPv4.InternalMasqueradeSubnet != "" {
+			if !cnHasIPv4 {
+				out = append(out, errors.Errorf("v4InternalMasqueradeSubnet %s and ClusterNetwork must have matching IP families", oc.GatewayConfig.IPv4.InternalMasqueradeSubnet))
+			}
+			if err := validateOVNKubernetesSubnet("v4InternalMasqueradeSubnet", oc.GatewayConfig.IPv4.InternalMasqueradeSubnet, &pool, conf.ClusterNetwork); err != nil {
+				out = append(out, err)
+			}
+		}
+		if oc.GatewayConfig.IPv6.InternalMasqueradeSubnet != "" {
+			if !cnHasIPv6 {
+				out = append(out, errors.Errorf("v6InternalMasqueradeSubnet %s and ClusterNetwork must have matching IP families", oc.GatewayConfig.IPv6.InternalMasqueradeSubnet))
+			}
+			if err := validateOVNKubernetesSubnet("v6InternalMasqueradeSubnet", oc.GatewayConfig.IPv6.InternalMasqueradeSubnet, &pool, conf.ClusterNetwork); err != nil {
+				out = append(out, err)
+			}
+		}
+	}
+
+	return kerrors.NewAggregate(out)
+}
+
+// Check subnet length, IP family and overlapping with other subnets
+func validateOVNKubernetesSubnet(name, subnet string, otherSubnets *iputil.IPPool, cn []operv1.ClusterNetworkEntry) error {
+	out := []error{}
+	_, cidr, err := net.ParseCIDR(subnet)
+	if err != nil {
+		out = append(out, errors.Errorf("%s is invalid: %s", name, err))
+	} else if name != "v4InternalMasqueradeSubnet" && !utilnet.IsIPv6CIDRString(subnet) && !isV4NodeSubnetLargeEnough(cn, subnet) {
+		out = append(out, errors.Errorf("%s %s is not large enough for the maximum number of nodes which can be supported by ClusterNetwork", name, subnet))
+	} else if name != "v6InternalMasqueradeSubnet" && utilnet.IsIPv6CIDRString(subnet) && !isV6NodeSubnetLargeEnough(cn, subnet) {
+		out = append(out, errors.Errorf("%s %s is not large enough for the maximum number of nodes which can be supported by ClusterNetwork", name, subnet))
+	}
```

Comment:
I don't understand. You are not actually aggregating the errors, you just return the first error you encounter.

This is fine but no need for append or anything and you can follow a simple `if ... { return fmt.Errorf(...) }` strategy.

Also, for the subnets that not need a size check, you can pass a nil `cn` to signal that it is not needed. If properly documented, is probably more maintainable than having to check for specific names which are just passed through for error formatting

Reply:
I thought adding all anomalies during the validation stage is a good idea and hence kept on appending errors instead of returning immediately upon encountering an error. Returning immediately also seems reasonable though.

I can start sending cn as nil for masquerade subnet.

If we dont pass name while formatting error then wont it be little bit confusing for customers? However I do agree that as per checks in place we are not allowing any overlaps with any configurable subnets. so, chances of any confusion is less.

Reply:
Yeah, continue passing the name to log it into the error, but do not use it to treat exceptions.